java/navigate-route/src/main/java/com/esri/arcgisruntime/sample/navigateroute/MainActivity.java

```diff
@@ -0,0 +1,280 @@
+import com.esri.arcgisruntime.mapping.view.MapView;
+import com.esri.arcgisruntime.navigation.DestinationStatus;
+import com.esri.arcgisruntime.navigation.RouteTracker;
+import com.esri.arcgisruntime.navigation.TrackingStatus;
+import com.esri.arcgisruntime.symbology.SimpleLineSymbol;
+import com.esri.arcgisruntime.tasks.networkanalysis.RouteParameters;
+import com.esri.arcgisruntime.tasks.networkanalysis.RouteResult;
+import com.esri.arcgisruntime.tasks.networkanalysis.RouteTask;
+import com.esri.arcgisruntime.tasks.networkanalysis.Stop;
+
+public class MainActivity extends AppCompatActivity {
+
+  private static final String TAG = MainActivity.class.getSimpleName();
+
+  private TextToSpeech mTextToSpeech;
+  private boolean mIsTextToSpeechInitialized = false;
+
+  private SimulatedLocationDataSource mSimulatedLocationDataSource;
+
+  private MapView mMapView;
+  private RouteTracker mRouteTracker;
+  private Graphic mRouteAheadGraphic;
+  private Graphic mRouteTraveledGraphic;
+  private Button mRecenterButton;
+
+  @Override
+  protected void onCreate(Bundle savedInstanceState) {
+    super.onCreate(savedInstanceState);
+    setContentView(R.layout.activity_main);
+
+    // get a reference to the map view
+    mMapView = findViewById(R.id.mapView);
+    // create a map and set it to the map view
+    ArcGISMap map = new ArcGISMap(Basemap.createStreetsVector());
+    mMapView.setMap(map);
+
+    // create a graphics overlay to hold our route graphics
+    GraphicsOverlay graphicsOverlay = new GraphicsOverlay();
+    mMapView.getGraphicsOverlays().add(graphicsOverlay);
+
+    // initialize text-to-speech to replay navigation voice guidance
+    mTextToSpeech = new TextToSpeech(this, status -> {
+      if (status != TextToSpeech.ERROR) {
+        mTextToSpeech.setLanguage(Resources.getSystem().getConfiguration().locale);
+        mIsTextToSpeechInitialized = true;
+      }
+    });
+
+    // clear any graphics from the current graphics overlay
+    mMapView.getGraphicsOverlays().get(0).getGraphics().clear();
+
+    // generate a route with directions and stops for navigation
+    RouteTask routeTask = new RouteTask(this, getString(R.string.routing_service_url));
+    ListenableFuture<RouteParameters> routeParametersFuture = routeTask.createDefaultParametersAsync();
+    routeParametersFuture.addDoneListener(() -> {
+
+      try {
+        // define the route parameters
+        RouteParameters routeParameters = routeParametersFuture.get();
+        routeParameters.setStops(getStops());
+        routeParameters.setReturnDirections(true);
+        routeParameters.setReturnStops(true);
+        routeParameters.setReturnRoutes(true);
+        ListenableFuture<RouteResult> routeResultFuture = routeTask.solveRouteAsync(routeParameters);
+        routeParametersFuture.addDoneListener(() -> {
+          try {
+            // get the route geometry from the route result
+            RouteResult routeResult = routeResultFuture.get();
```

Comment:
It seems that you are doing things twice: getting the geometry, creating the graphic for the route, and setting it to the graphics overlay, as this is done here and inside the function `startNavigation`. `startNavigation` gets called in line 127 so is there any benefit for doing these things twice?

Reply:
Ahh I think I see what you mean, but one is inside the onclick listener and one is not. So basically:

- startNavigation is called once, once a route as been calculated
- startNavigation is called again whenever the "Navigate Route" button is pressed

Reply:
@TADraeseke Thanks for clarifying.